server/expression.js

```diff
@@ -378,80 +410,109 @@ expression.listMatrix_ = function(expmatAddr) {
-  }
   var lines = fs.readFileSync(expressionFile).toString().split('\n');
   lines.forEach(function(line) {
     var parts = line.split('\t');
     if (isFirstCol) {
-      // first column contains the conditions
+      // first row contains the conditions
       isFirstCol = false;
-      for (var i = 1; i < parts.length; i++) {
-        if (parts[i].match(expCondition)) {
-          conditions.push(i);
-          validCol.push(true);
-          conditionNames.push(parts[i]);
-        } else {
-          validCol.push(false);
+      conditionNames.forEach(function(conditionName) {
+        var conditionIndex = parts.indexOf(conditionName);
+        if (conditionIndex != -1 && conditionIndex != 0) {
+          conditions.push(conditionIndex);
         }
+      });
+      for (var i = 1; i < parts.length; i++) {
+        allConditionNames.push(parts[i]);
       }
     } else {
-      // other columns contain a gene, and values
-      if (parts[0].match(expGene)) {
-        geneNames.push(parts[0]);
+      // other rows contain a gene, and values
+      var geneIndex = geneNames.indexOf(parts[0]);
+      if (geneIndex != -1) {
         var tmpLine = [];
-        for (var i = 1; i < parts.length; i++) {
-          var value = parseFloat(parts[i]);
-          if (validCol[i - 1]) {
-            valueMin = Math.min(valueMin, value);
-            valueMax = Math.max(valueMax, value);
-            tmpLine.push(value);
-          }
-          allValueMin = Math.min(allValueMin, value);
-          allValueMax = Math.max(allValueMax, value);
-        }
-        values.push(tmpLine);
+        conditions.forEach(function(conditionIndex) {
+          var value = parseFloat(parts[conditionIndex]);
+          valueMin = Math.min(valueMin, value);
+          valueMax = Math.max(valueMax, value);
+          tmpLine.push(value);
+        });
+        values[parts[0]] = tmpLine;
       }
-      else {
-        for (var i = 1; i < parts.length; i++) {
-          var value = parseFloat(parts[i]);
-          allValueMin = Math.min(allValueMin, value);
-          allValueMax = Math.max(allValueMax, value);
-        }
+      allGeneNames.push(parts[0]);
+      for (var i = 1; i < parts.length; i++) {
+        var value = parseFloat(parts[i]);
+        allValueMin = Math.min(allValueMin, value);
+        allValueMax = Math.max(allValueMax, value);
       }
     }
   });
+  Object.keys(values)
+    .sort()
+    .forEach(function(geneName, i) {
+      sortedValues.push(values[geneName]);
+    });
   return {
-    values: values,
+    values: sortedValues,
     geneNames: geneNames,
     conditionNames: conditionNames,
     valueMin: valueMin,
-    valueMax: valueMax,
+    valueMax: valueMax
+  };
+};
+
+/**
+ * Reads all of the expression matrix data from text file.
+ * @param {string} expressionFile Path to the expression file.
+ * @return {expression.Matrix}
+ * @private
+ */
+expression.readExpressionAll_ = function(expressionFile) {
```

Comment:
Combine this with readExpression if possible, using an extra flag. If not possible, then give it a clearer name, like getMatrixInfo.